### PR TITLE
We are going to need a bigger Bytes

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: base/array.hpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+namespace principia {
+namespace base {
+
+// A simple container for a pointer and size, similar to a StringPiece.  |data|
```
Comment (eggrobin) [inline base/array.hpp]: Not really anymore (maybe make that comment on Bytes)

Comment (eggrobin) [inline base/array.hpp]: Or just drop it.

File: base/array.hpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+           typename = std::enable_if_t<std::is_integral<Size>::value, Size>>
+  explicit UniqueArray(Size const size);
+  // Takes ownership of an existing array.
+  template<typename Size,
+           typename = std::enable_if_t<std::is_integral<Size>::value, Size>>
+  UniqueArray(std::unique_ptr<Element[]> data, Size const size);
+
+  // Move it, move it!
+  UniqueArray(UniqueArray&& other);
+  UniqueArray& operator=(UniqueArray&& other);
```
Comment (eggrobin) [inline base/array.hpp]: Comment: should be defaulted when the compiler allows that.

File: base/hexadecimal.hpp [hunk 3]
```diff
@@ -1,26 +1,30 @@
...
-// |input_size|.  The range [&output[input_size << 1], &output[output_size])
-// is left unmodified.
-void HexadecimalEncode(uint8_t const* input, int64_t const input_size,
-                       uint8_t* output, int64_t const output_size);
+// The result is upper-case.  Either |input.data <= &output->data[1]| or
+// |&output->data[input.size << 1] <= input.data| must hold, in particular,
+// |input.data == output->data| is valid.  |output.size| must be at least twice
+// |input.size|.  The range
+// [&output->data[input.size << 1], &output->data[output->size][ is left
+// unmodified.
```
Comment (eggrobin) [inline base/hexadecimal.hpp]: arrows

File: base/hexadecimal_body.hpp [hunk 4]
```diff
@@ -34,43 +35,47 @@ static uint8_t const kHexadecimalDigitsToNibble[256] = {
...
-    *output = (kHexadecimalDigitsToNibble[*input] << 4) |
-              kHexadecimalDigitsToNibble[*(input + 1)];
+  CHECK(output.data <= &input.data[1] ||
+        &input.data[input.size] <= output.data) << "bad overlap";
+  CHECK_GE(output.size, input.size / 2) << "output too small";
+  for (uint8_t const* const input_end = input.data + input.size;
+       input.data != input_end;
+       input.data += 2, ++output.data) {
+    *output.data = (kHexadecimalDigitsToNibble[*input.data] << 4) |
+                     kHexadecimalDigitsToNibble[*(input.data + 1)];
```
Comment (eggrobin) [inline base/hexadecimal_body.hpp]: Bad indent.

File: base/array.hpp (hunk 1 above)
Comment (pleroy) [inline base/array.hpp]: Done.

File: base/array.hpp (hunk 2 above)
Comment (pleroy) [inline base/array.hpp]: Done.

File: base/hexadecimal.hpp (hunk 3 above)
Comment (pleroy) [inline base/hexadecimal.hpp]: Done.

File: base/hexadecimal_body.hpp (hunk 4 above)
Comment (pleroy) [inline base/hexadecimal_body.hpp]: Done.

**Pull request merged**